File: src/main/buttons/Group.java
package main.buttons;

import processing.core.PApplet;
import processing.core.PImage;
import processing.data.JSONArray;
import processing.data.JSONObject;

import java.io.File;
import java.util.ArrayList;
import java.util.HashSet;

public class Group extends Button implements Comparable<Group> {

    public static final int SIZE = 128;
    public static final int GAP = 30;
    public static final int GROUP_X = GAP;
    public static final int GROUP_Y = GAP + 20;
    private static final float GROUP_MOVING_RATE = 15;
    private static final int ALPHA_CHANGE = 20;

    public static int groupCountX;
    public static int groupCountY;
    private static int maxGroups;

    public static int totalPages;
    public static int pageNumber = 0;

    static int groupSelectedX;
    public static int groupSelectedAY;
    public static int groupSelectedBY;

    public static Group groupSelectedA;
    public static Group groupSelectedB;

    private static float deltaX;
    private static float deltaY;

    private static boolean moving = false;

    private int colour;
    private int alpha = 255;
    private int alphaChange;
    private boolean done = false;

    private String name;
    private Pack pack;

    private Group(JSONObject json, Pack pack) {
        super(SIZE, SIZE);

        this.pack = pack;
        this.name = pack.getNamespacedName(json.getString("name"));
        JSONArray colourArray = json.getJSONArray("colour");
        this.colour = main.color(colourArray.getInt(0), colourArray.getInt(1), colourArray.getInt(2));

        //check if a pack has the image, from top to bottom
        for (Pack pack1 : main.packsRoom.getLoadedPacks()) {
            if (pack1.getName().equals("Alchemy") && this.pack.getName().equals("Alchemy")) {
                //if the element is of the default pack and we are in the default pack right now, load default location
                this.setImage(main.loadImage("resources/groups/alchemy/" + this.getID() + ".png"));
                break;
            } else {
                String packPath = pack1.getPath() + "/groups/" + this.getPack().getNamespace() + "/" + this.getID() + ".png";
                if (new File(packPath).exists()) {
                    this.setImage(main.loadImage(packPath));
                    break;
                }
            }
        }
        this.getImage().resize(SIZE, SIZE);
    }

    //copy constructor
    private Group(String name, PImage image, Pack pack, int colour, int alpha, int alphaChange, boolean done) {
        super(SIZE, SIZE, image);

        this.name = name;
        this.pack = pack;
        this.colour = colour;
        this.alpha = alpha;
        this.alphaChange = alphaChange;
        this.done = done;
    }

    public String getName() {
        return this.name;
    }

    Pack getPack() {
        return this.pack;
    }

    String getID() {
        return this.name.split(":")[1];
    }

    //just added an extra !moving condition
    @Override
    protected boolean inBounds() {
        return main.mouseX >= this.getX() && main.mouseX <= this.getX() + SIZE && main.mouseY >= this.getY() && main.mouseY <= this.getY() + SIZE && !moving;
    }

    @Override
    public int compareTo(Group o) {
        if (!this.getID().equals(o.getID())) {
            return this.getID().compareTo(o.getID());
        } else {
            int thisPack = main.packsRoom.getLoadedPacks().indexOf(this.pack);
            int oPack = main.packsRoom.getLoadedPacks().indexOf(o.pack);
            return Integer.compare(thisPack, oPack);
        }
    }

    public static void reset() {
        pageNumber = 0;
        groupSelectedA = null;
        groupSelectedB = null;
    }

    public static void loadGroups(JSONArray array, Pack pack) {
        for (int i = 0; i < array.size(); i++) {
            JSONObject object = array.getJSONObject(i);
            if (object.hasKey("remove")) {
                if (object.getString("remove").equals("all")) {
                    main.loading.removeAllGroups();
                    main.groups.clear();
                } else {
                    Group group = getGroup(pack.getNamespacedName(object.getString("remove")));
                    if (group == null) {
                        System.err.println(pack.getNamespacedName(object.getString("remove")) + " group not found!");
                    } else {
                        main.loading.removeGroup(group);
                        main.groups.remove(group);
                    }
                }
            } else {
                main.groups.put(new Group(object, pack), new HashSet<>());
            }
            main.loading.updateProgress();
        }
    }

    public static Group getGroup(String name) {
        for (Group group : main.groups.keySet()) {
            if (group.name.equals(name)) {
                return group;
            }
        }
        return null;
    }

    public static void drawGroups() {
        int x = GROUP_X;
        int y = GROUP_Y;
        int maxX = (int) Math.round(main.screenWidth / 2F * 0.6); //maximum X value to draw the group grid

        groupSelectedX = maxX + 100;
        groupSelectedAY = y;
        groupSelectedBY = (int) Math.round(main.screenHeight * 0.44);

        int maxY = main.screenHeight - 60; //maximum Y value to draw the group grid
        //determine how many groups to draw horizontally
        groupCountX = Math.floorDiv(maxX - x, SIZE + GAP);
        //determine how many groups to draw vertically
        groupCountY = Math.floorDiv(maxY - y, SIZE + GAP);
        //number of groups on a page
        maxGroups = groupCountX * groupCountY;

        totalPages = Math.round(PApplet.ceil((float) main.game.getDiscovered().keySet().size() / maxGroups));

        ArrayList<Group> groups = getGroups();
        for (int i = 0; i < groups.size(); i++) {
            groups.get(i).draw(x, y);
            x += (SIZE + GAP);
            if ((i + 1) % groupCountX == 0) {
                x = GROUP_X;
                y += (SIZE + GAP);
            }
        }

        if (groupSelectedA != null) {
            if (groupSelectedA.getX() < groupSelectedX) {
                groupSelectedA.incrementX(deltaX / GROUP_MOVING_RATE);
                groupSelectedA.incrementY(deltaY / GROUP_MOVING_RATE);
                //having the if statement here ensures that it only gets run once
                if (groupSelectedA.getX() >= groupSelectedX) {
                    //need to set it so it doesn't go past
                    groupSelectedA.setX(groupSelectedX);
                    groupSelectedA.setY(groupSelectedAY);
                    moving = false;
                }
            }
            //needs to be constantly set due to screen resizing
            if (!moving) {
                groupSelectedA.setX(groupSelectedX);
                groupSelectedA.setY(groupSelectedAY);
            }
            groupSelectedA.draw();
            if (groupSelectedA.done) {
                groupSelectedA = null;
            }
        }

        if (groupSelectedB != null) {
            if (groupSelectedB.getX() < groupSelectedX) {
                groupSelectedB.incrementX(deltaX / GROUP_MOVING_RATE);
                groupSelectedB.incrementY(deltaY / GROUP_MOVING_RATE);
                //having the if statement here ensures that it only gets run once
                if (groupSelectedB.getX() >= groupSelectedX) {
                    groupSelectedB.setX(groupSelectedX);
                    groupSelectedB.setY(groupSelectedBY);
                    moving = false;
                }
            }
            //needs to be constantly set due to screen resizing
            if (!moving) {
                groupSelectedB.setX(groupSelectedX);
                groupSelectedB.setY(groupSelectedBY);
            }
            groupSelectedB.draw();
            if (groupSelectedB.done) {
                groupSelectedB = null;
            }
        }

        main.tint(255, 255);
    }

    @Override
    protected void drawButton() {
        this.updateAlpha();
        main.image(this.getImage(), this.getX(), this.getY());
    }

    private void updateAlpha() {
        if (this.alphaChange > 0) { //fade in
            if (this.alpha < 255) {
                this.alpha += this.alphaChange;
            }
            if (this.alpha >= 255) {
                this.alpha = 255;
            }
        } else if (this.alphaChange < 0) { //fade out
            if (this.alpha > 0) {
                this.alpha += this.alphaChange;
            }
            if (this.alpha <= 0) { //completely invisible
                this.alpha = 0;
                this.done = true;
            }
        }
        main.tint(255, this.alpha);
    }

    public static ArrayList<Group> getGroups() {
        ArrayList<Group> groupList = new ArrayList<>(main.game.getDiscovered().keySet());
        ArrayList<Group> list = new ArrayList<>(); //groups that are actually on screen
        for (int i = pageNumber * maxGroups; i < (pageNumber + 1) * maxGroups; i++) {
            if (i < groupList.size()) {
                list.add(groupList.get(i));
            }
        }
        //not a single group is drawn, page is empty, go to previous page
        if (list.size() == 0 && pageNumber > 0) {
            pageNumber--;
            return getGroups();
        }
        return list;
    }

    private boolean isSelected() {
        return this.getX() >= groupSelectedX;
    }

    int getColour() {
        return this.colour;
    }

    boolean exists() {
        for (Group group : main.game.getDiscovered().keySet()) {
            if (group.name.equals(this.name)) {
                return true;
            }
        }
        return false;
    }

    private Group deepCopy() {
        Group clone = new Group(this.name, this.getImage(), this.pack, this.colour, this.alpha, this.alphaChange, this.done);
        clone.setX(this.getX());
        clone.setY(this.getY());
        return clone;
    }

    @Override
    public void clicked() {
        if (!moving) {
            if (!this.isSelected()) {
                //the two blocks of code should be identical
                if (groupSelectedA == null) {
                    groupSelectedA = this.deepCopy();
                    moving = true;
                    deltaX = groupSelectedX - groupSelectedA.getX();
                    deltaY = groupSelectedAY - groupSelectedA.getY();
                    groupSelectedA.alpha = 0; //because by default alpha is 255
                    groupSelectedA.alphaChange = ALPHA_CHANGE;
                    Element.resetA();
                } else if (groupSelectedB == null) {
                    groupSelectedB = this.deepCopy();
                    moving = true;
                    deltaX = groupSelectedX - groupSelectedB.getX();
                    deltaY = groupSelectedBY - groupSelectedB.getY();
                    groupSelectedB.alpha = 0; //because by default alpha is 255
                    groupSelectedB.alphaChange = ALPHA_CHANGE;
                    Element.resetB();
                }
            } else {
                if (this == groupSelectedA) {
                    groupSelectedA.alphaChange = -ALPHA_CHANGE;
                    Element.hidePagesA();
                } else if (this == groupSelectedB) {
                    groupSelectedB.alphaChange = -ALPHA_CHANGE;
                    Element.hidePagesB();
                }
            }
        }
    }

    public static void setHintGroups(Group a, Group b) {
        groupSelectedA = a.deepCopy();
        groupSelectedB = b.deepCopy();
        Element.resetA();
        Element.resetB();
    }

}

File: src/main/buttons/Group.kt
package main.buttons

import main.Element
import main.Main
import main.buttons.ElementButton.Companion.hidePagesA
import main.buttons.ElementButton.Companion.hidePagesB
import main.buttons.ElementButton.Companion.resetA
import main.buttons.ElementButton.Companion.resetB
import main.rooms.Game
import main.rooms.Game.discovered
import main.rooms.GameMode
import main.rooms.Loading.removeAllGroups
import main.rooms.Loading.removeGroup
import main.rooms.Loading.updateProgress
import main.rooms.PacksRoom.loadedPacks
import processing.core.PImage
import processing.data.JSONArray
import processing.data.JSONObject
import java.io.File
import java.util.*
import kotlin.math.ceil
import kotlin.math.min
import kotlin.math.roundToInt

class Group(
    val id: String,
    val colour: Int,
    image: PImage,
    val pack: Pack,
    private var alpha: Int = 255,
    private var alphaChange: Int = 0
) : Button(SIZE, SIZE, image), Comparable<Group> {

    private var done = false
    val localId = getLocalId(id)

    private fun isSelected() = x >= groupSelectedX

    //just added an extra !moving condition
    override fun inBounds(): Boolean {
        return Main.mouseX >= x && Main.mouseX <= x + SIZE && Main.mouseY >= y && Main.mouseY <= y + SIZE && !moving
    }

    override fun compareTo(other: Group): Int {
        return if (localId != other.localId) {
            localId.compareTo(other.localId)
        } else {
            val thisPack = loadedPacks.indexOf(pack)
            val oPack = loadedPacks.indexOf(other.pack)
            thisPack.compareTo(oPack)
        }
    }

    override fun drawButton() {
        updateAlpha()
        Main.image(image, x, y)
    }

    private fun updateAlpha() {
        if (alphaChange > 0) { //fade in
            if (alpha < 255) {
                alpha += alphaChange
            }
            if (alpha >= 255) {
                alpha = 255
            }
        } else if (alphaChange < 0) { //fade out
            if (alpha > 0) {
                alpha += alphaChange
            }
            if (alpha <= 0) { //completely invisible
                alpha = 0
                done = true
            }
        }
        Main.tint(255, alpha.toFloat())
    }

    fun exists() = discovered.keys.any { it.id == id }

    override fun clicked() {
        if (!moving) {
            if (!isSelected()) {
                //in puzzle mode, you cannot open the same group twice
                if (Game.mode == GameMode.PUZZLE && (this.id == groupSelectedA?.id || this.id == groupSelectedB?.id)) {
                    return
                }
                //the two blocks of code should be identical
                if (groupSelectedA == null) {
                    groupSelectedA = Group(id, colour, image, pack, 0, ALPHA_CHANGE)
                    groupSelectedA!!.x = x
                    groupSelectedA!!.y = y
                    moving = true
                    deltaX = groupSelectedX - groupSelectedA!!.x
                    deltaY = groupSelectedAY - groupSelectedA!!.y
                    resetA()
                } else if (groupSelectedB == null) {
                    groupSelectedB = Group(id, colour, image, pack, 0, ALPHA_CHANGE)
                    groupSelectedB!!.x = x
                    groupSelectedB!!.y = y
                    moving = true
                    deltaX = groupSelectedX - groupSelectedB!!.x
                    deltaY = groupSelectedBY - groupSelectedB!!.y
                    resetB()
                }
            } else {
                if (this === groupSelectedA) {
                    groupSelectedA?.alphaChange = -ALPHA_CHANGE
                    hidePagesA()
                } else if (this === groupSelectedB) {
                    groupSelectedB?.alphaChange = -ALPHA_CHANGE
                    hidePagesB()
                }
            }
        }
    }

    companion object {
        const val SIZE = 128
        const val GAP = 30
        const val GROUP_X: Int = GAP
        const val GROUP_Y: Int = GAP + 20
        private const val GROUP_MOVING_RATE = 15f
        private const val ALPHA_CHANGE = 20
        var groupCountX = 0
        var groupCountY = 0
        private var maxGroups = 0
        var totalPages = 0
        var pageNumber = 0
        var groupSelectedX = 0
        var groupSelectedAY = 0
        var groupSelectedBY = 0
        var groupSelectedA: Group? = null
        var groupSelectedB: Group? = null
        private var deltaX = 0f
        private var deltaY = 0f
        private var moving = false

        fun reset() {
            pageNumber = 0
            groupSelectedA = null
            groupSelectedB = null
        }

        fun loadGroups(array: JSONArray, pack: Pack) {
            for (i in 0 until array.size()) {
                val json = array.getJSONObject(i)
                if (json.hasKey("remove")) {
                    if (json.getString("remove") == "all") {
                        removeAllGroups()
                        Main.groups.clear()
                    } else {
                        val group = getGroup(pack.getNamespacedName(json.getString("remove")))
                        if (group == null) {
                            System.err.println(pack.getNamespacedName(json.getString("remove")) + " group not found!")
                        } else {
                            removeGroup(group)
                            Main.groups.remove(group)
                        }
                    }
                } else {
                    Main.groups[createGroup(json, pack)] = HashSet<Element>()
                }
                updateProgress()
//                Thread.sleep(500)
            }
        }

        private fun getLocalId(id: String) = id.split(":")[1]

        private fun createGroup(json: JSONObject, pack: Pack): Group {
            val id = pack.getNamespacedName(json.getString("name"))
            val colourArray = json.getJSONArray("colour")
            lateinit var image: PImage

            //check if a pack has the image, from top to bottom
            for (loadedPack in loadedPacks) {
                if (loadedPack.name == "Alchemy" && pack.name == "Alchemy") {
                    //if the element is of the default pack and we are in the default pack right now, load default location
                    image = Main.loadImage("resources/groups/alchemy/${getLocalId(id)}.png")
                    break
                } else {
                    val packPath = "${loadedPack.path}/groups/${pack.namespace}/${getLocalId(id)}.png"
                    if (File(packPath).exists()) {
                        image = Main.loadImage(packPath)
                        break
                    }
                }
            }
            return Group(
                id,
                Main.color(colourArray.getInt(0), colourArray.getInt(1), colourArray.getInt(2)),
                image,
                pack
            )
        }

        @JvmStatic
        fun getGroup(id: String): Group? = Main.groups.keys.find { it.id == id }

        fun drawGroups() {
            var x = GROUP_X
            var y = GROUP_Y
            val maxX = (Main.screenWidth / 2f * 0.6).roundToInt() //maximum X value to draw the group grid
            groupSelectedX = maxX + 100
            groupSelectedAY = y
            groupSelectedBY = (Main.screenHeight * 0.44).roundToInt()
            val maxY: Int = Main.screenHeight - 60 //maximum Y value to draw the group grid
            //determine how many groups to draw horizontally
            groupCountX = Math.floorDiv(maxX - x, SIZE + GAP)
            //determine how many groups to draw vertically
            groupCountY = Math.floorDiv(maxY - y, SIZE + GAP)
            //number of groups on a page
            maxGroups = groupCountX * groupCountY
            totalPages = ceil(discovered.keys.size.toFloat() / maxGroups).roundToInt()
            if (pageNumber >= totalPages) {
                pageNumber = totalPages - 1
            }
            for ((index, group) in getCurrentPageGroups().withIndex()) {
                group.draw(x.toFloat(), y.toFloat())
                x += SIZE + GAP
                if ((index + 1) % groupCountX == 0) {
                    x = GROUP_X
                    y += SIZE + GAP
                }
            }
            groupSelectedA?.let {
                if (it.x < groupSelectedX) {
                    it.x += deltaX / GROUP_MOVING_RATE
                    it.y += deltaY / GROUP_MOVING_RATE
                    //having the if statement here ensures that it only gets run once
                    if (it.x >= groupSelectedX) {
                        //need to set it so it doesn't go past
                        it.x = groupSelectedX.toFloat()
                        it.y = groupSelectedAY.toFloat()
                        moving = false
                    }
                }
                //needs to be constantly set due to screen resizing
                if (!moving) {
                    it.x = groupSelectedX.toFloat()
                    it.y = groupSelectedAY.toFloat()
                }
                it.draw()
            }
            if (groupSelectedA?.done == true) {
                groupSelectedA = null
            }
            groupSelectedB?.let {
                if (it.x < groupSelectedX) {
                    it.x += deltaX / GROUP_MOVING_RATE
                    it.y += deltaY / GROUP_MOVING_RATE
                    //having the if statement here ensures that it only gets run once
                    if (it.x >= groupSelectedX) {
                        it.x = groupSelectedX.toFloat()
                        it.y = groupSelectedBY.toFloat()
                        moving = false
                    }
                }
                //needs to be constantly set due to screen resizing
                if (!moving) {
                    it.x = groupSelectedX.toFloat()
                    it.y = groupSelectedBY.toFloat()
                }
                it.draw()
            }
            if (groupSelectedB?.done == true) {
                groupSelectedB = null
            }
            Main.tint(255, 255f)
        }

        fun setHintGroups(a: Group, b: Group) {
            groupSelectedA = Group(a.id, a.colour, a.image, a.pack)
            groupSelectedB = Group(b.id, b.colour, b.image, b.pack)
            resetA()
            resetB()
        }

        fun getCurrentPageGroups(): List<Group> {
            return discovered.keys.toList()
                .subList(pageNumber * maxGroups, min((pageNumber + 1) * maxGroups, discovered.keys.size))
        }
    }
}